Make all components optional

nmittler opened the pull request:
The DNS proxy is already optional. This PR also makes the HBONE proxy optional, which allows running ztunnel as a stand-alone DNS proxy.

Moved the DNS proxy to a top-level package.

This also cleans up the metrics/stats and readiness servers to make the code more consistent.

Fixes #581

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | d1c4888c4882650c36d8b71efeb3784cf4f82909 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1678083118549962752) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | 8a0d37339dbafed2953b1bc43126e03170c53615 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1678087664256946176) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
Added a hold on this until https://github.com/istio/ztunnel/pull/597 lands

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | 0be097027ef5043492d39cb710dce6fef1efbcc3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1678891588513697792) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | 10937153a23a322c96db5096b424f35330f77c74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1678893813197705216) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | 341a4ee02c6fe7ccbaa1a131d67232b5c7b287b5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1678895795941675008) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | dee53c0bc73bb13c6f05e0aa7f93abad9d4ee164 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1678898116557803520) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | 8451b8668295f6ef97dbdd63b89e7ff61b739ce4 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1678902685417345024) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
@howardjohn mind taking another pass? I refactored the startup app flow a bit more and separated the dns metrics out of the global metrics struct.

howardjohn commented on src/app.rs:
```diff
@@ -12,78 +12,133 @@
+        .await
+        .context("readiness server starts")?;
+    let readiness_address = readiness_server.address();
+    readiness_server.spawn();
```
I think we need this in the data plane task pool, as discussed on slack

nmittler replied:
My takeaway from our discussion is that this one was on the edge of control/data plane and didn't need to be in the data plane task pool?

howardjohn commented on src/metrics.rs:
```diff
@@ -83,57 +66,70 @@ impl Metrics {
     }
 }
 
+pub trait DeferRecorder {
```
meant for another PR?

nmittler replied:
I had to extract this from `Metrics` because I separated out the DNS metrics. Now both structures implement this.

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | 50168a9801cee7fe69f666570a9f72dbeeafbd76 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1679197966801309696) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | 263d759e295b07b4a7f881474eea3e0fc98d513e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1679219836506345472) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@nmittler: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
test_ztunnel | a0080e99af1a6570b93a87db546c8fe65967fec1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_ztunnel/592/test_ztunnel/1679226325862191104) | true | `/test test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

nmittler commented:
@howardjohn this is ready to go ... PTAL.

I've refactored things a bit further. Metrics are now completely split out into the pieces that are needed by each component. I've also reworked the `app`'s data plane work pool and use a channel to add tasks to the data plane pool. We now eagarly start the readiness server (in the dataplane pool), so that we can properly serve readiness for all the other components until they're actually ready to serve.
